File: app/src/main/java/com/porarrirr/sumahohikakuku/domain/input/InputSanitizer.kt
package com.porarrirr.sumahohikakuku.domain.input

import java.util.Locale

fun sanitizeDecimalInput(raw: String): String {
    if (raw.isEmpty()) return ""
    val normalized = buildString(raw.length) {
        for (char in raw) {
            val ascii = when (char) {
                in '０'..'９' -> (char.code - '０'.code + '0'.code).toChar()
                '．', '。', '，', ',' -> '.'
                else -> char
            }
            if (ascii.isDigit() || ascii == '.') append(ascii)
        }
    }
    val firstDot = normalized.indexOf('.')
    if (firstDot == -1) return normalized
    return normalized.substring(0, firstDot + 1) + normalized.substring(firstDot + 1).filter { it != '.' }
}

fun sanitizeHexInput(raw: String): String {
    if (raw.isEmpty()) return "#"
    val cleaned = raw.uppercase(Locale.US)
        .replace("#", "")
        .filter(::isHexDigit)
        .take(6)
    return if (cleaned.isEmpty()) "#" else "#$cleaned"
}

fun parseHexColor(input: String): String? {
    val cleaned = input.uppercase(Locale.US).removePrefix("#")
    return if (cleaned.length == 6 && cleaned.all(::isHexDigit)) "#$cleaned" else null
}

private fun isHexDigit(char: Char): Boolean {
    return char in '0'..'9' || char in 'A'..'F'
}

File: ios/src/Domain/InputSanitizer.swift
import Foundation

public func sanitizeDecimalInput(raw: String) -> String {
    if raw.isEmpty { return "" }
    var normalized = ""
    for char in raw {
        let ascii: Character
        switch char {
        case "０"..."９":
            let code = Int(char.unicodeScalars.first!.value) - Int(Character("０").unicodeScalars.first!.value) + Int(Character("0").unicodeScalars.first!.value)
            ascii = Character(UnicodeScalar(code)!)
        case "．", "。", "，", ",":
            ascii = "."
        default:
            ascii = char
        }
        if ascii.isNumber || ascii == "." {
            normalized.append(ascii)
        }
    }
    guard let firstDot = normalized.firstIndex(of: ".") else {
        return normalized
    }
    let beforeDot = normalized[..<firstDot]
    let afterDot = normalized[firstDot...].filter { $0 != "." }
    return String(beforeDot) + "." + String(afterDot)
}

public func sanitizeHexInput(raw: String) -> String {
    if raw.isEmpty { return "#" }
    let cleaned = raw.uppercased()
        .replacingOccurrences(of: "#", with: "")
        .filter { isHexDigit($0) }
        .prefix(6)
    return cleaned.isEmpty ? "#" : "#\(cleaned)"
}

public func parseHexColor(input: String) -> String? {
    let cleaned = input.uppercased().trimmingCharacters(in: CharacterSet(charactersIn: "#"))
    return (cleaned.count == 6 && cleaned.allSatisfy { isHexDigit($0) }) ? "#\(cleaned)" : nil
}

private func isHexDigit(_ char: Character) -> Bool {
    return char.isNumber || ("A"..."F").contains(char)
}
